allow to set identity_func by ANALYTICAL_IDENTITY_FUNC settings

PetrDlouhy opened the pull request:
Google has in it's conditions for enabling UserID requirement, that prohibits sending personal data (such as e-mail address) to analytics.
I am using e-mail address as username, so using GTag would break the requirements for me.

This enables me to set user UUID for the UserID parameter.

bittner reviewed (changes requested):
Can you extend the tests to match your changes, please? And also mention `ANALYTICAL_IDENTIFY_FUNC` in the docs.

Thank you for your understanding.

bittner commented on analytical/utils.py:
```diff
@@ -84,6 +84,8 @@ def get_identity(context, prefix=None, identity_func=None, user=None):
+                                        identity_func)
                 if identity_func is not None:
                     return identity_func(user)
                 else:
```
I know this is not part of the PR, but if you wouldn't mind: While you're here touching the code, I'd like to revert the `if`-clause to reduce cognitive overhead, i.e.

```python
                if identity_func is None:
                    return user.get_username()
                else:
                    return identity_func(user)
```
Feel free to ignore this request.

PetrDlouhy commented:
@bittner I will look at it.

PetrDlouhy commented:
@bittner I added simple test, some docs and also renamed the setting from ANALITICAL_IDENTIFY_FUNC to ANALYTICAL_IDENTITY_FUNC.
Please approve the workflows.

bittner commented on docs/settings.rst:
```diff
@@ -25,3 +25,17 @@ their default values.
+
+.. data:: ANALYTICAL_IDENTITY_FUNC
+
+    Default: Identity function dependend on provider
```
```suggestion
    Default: Identity function dependent on provider
```

bittner commented on docs/settings.rst:
```diff
@@ -25,3 +25,17 @@ their default values.
+    Default: Identity function dependend on provider
+
+    A function that returns the identity of the given user. This overrides the
+    default settings of diferent providers.
```
```suggestion
    default settings of different providers.
```

bittner commented on docs/settings.rst:
```diff
@@ -25,3 +25,17 @@ their default values.
+    A function that returns the identity of the given user. This overrides the
+    default settings of diferent providers.
+
+    E.g. Google has in it's conditions for enabling UserID requirement, that prohibits
```
```suggestion
    E.g. Google has in its conditions for enabling UserID the requirement, that prohibits
```

bittner commented on docs/settings.rst:
```diff
@@ -25,3 +25,17 @@ their default values.
+    default settings of diferent providers.
+
+    E.g. Google has in it's conditions for enabling UserID requirement, that prohibits
+    sending personal data (such as e-mail address) to analytics.
```
```suggestion
    sending personal data (such as an e-mail address) to analytics.
```

bittner reviewed (changes requested):
IIUC, the problem you're trying to solve relates to hard requirements of some of Analytical's analytics providers. Shouldn't this be solved in the respective modules instead of requiring our users to code an identity function?

PetrDlouhy commented:
@bittner I have fixed the typos.

I think, that possibility to manually override the identity function allows flexibility in corner cases. E.g. I want to use my UUID field to keep consistency of my data even if the default behavior of `gtag` provider is fixed.

I don't know other providers, but the `django-analytical` might use hash from the username as identity function for `google_analytics_gtag` provider. But that would interrupt consistency of the collected data for users that are updating this app. So I am not sure how to update this correctly (Make a new provider and mark the current obsolete?).

It is also truth, that one identity function might not fit for all providers, if more of them are used. So maybe we might better use settings like:
```python
ANALYTICAL_IDENTITY_FUNC_OVERRIDES = {
    'google_analytics_gtag': lambda user: user.uuid,
}
```
and
```python
ANALYTICAL_IDENTITY_FUNC_DEFAULT = lambda user: user.uuid
```
which would override only the default `user.get_username()` function.

bittner reviewed (changes requested):
I agree that making the identity func dependent on the provider would make sense. I would keep it simple, though.

```python
ANALYTICAL_IDENTITY_FUNC = {
    'google_analytics_gtag': lambda user: user.uuid,
}
```

or maybe even

```python
ANALYTICAL_AUTO_IDENTIFY = {
    'google_analytics_gtag': lambda user: user.uuid,
    'matomo': True,
}
```

Though I just noticed something else: The feature you attempt to contribute already exists, doesn't it? Instead of a (lamda) function, though, you simply add an `analytical_identity` value (or the service specific value) to the template context. – Isn't that good enough?

PetrDlouhy commented on docs/settings.rst:
```diff
@@ -25,3 +25,17 @@ their default values.
+    A function that returns the identity of the given user. This overrides the
+    default settings of different providers.
+
+    E.g. Google has in its conditions for enabling UserID the requirement, that prohibits
+    sending personal data (such as an e-mail address) to analytics.
+    If e-mail address is used as username, using GTag would break the requirements.
```
I meant transition from `google_analytics` to `google_analytics_gtag` with user identity which I underwent in my project.

The requirements are their terms of use, which they write very suggestively: https://developers.google.com/analytics/solutions/crm-integration#user_id
And also I think I had to agree to that when switching on the User ID feature in Google Analytics.

bittner commented on analytical/utils.py:
```diff
@@ -79,15 +79,17 @@ def get_identity(context, prefix=None, identity_func=None, user=None):
     except KeyError:
         pass
-    if getattr(settings, 'ANALYTICAL_AUTO_IDENTIFY', True):
+    if getattr(settings, 'ANALYTICAL_AUTO_IDENTITY', True):
```
It's in [the documentation](https://django-analytical.readthedocs.io/en/latest/features.html#identifying-visitors) like that.

PetrDlouhy replied:
Yeah, sorry for that. I fixed it even thou we might not use this PR.

PetrDlouhy commented:
You are right. The `analytical_identity` in context works. Maybe we should document it better (it is described under few providers, but not in general docs nor for `google_analytics_gtag` provider.

Also some providers (like `google_analytics_gtag`) doesn't pass the `prefix` parameter, so it is not possible to use provider specific value.

bittner commented:
> Maybe we should document it better (it is described under few providers, but not in general docs nor for `google_analytics_gtag` provider.

Great! It would be awesome if you could provide the missing pieces to the documentation. A working code sample would be fantastic.

> Also some providers (like `google_analytics_gtag`) doesn't pass the `prefix` parameter, so it is not possible to use provider specific value.

That was already reported (#188, #197). Can you spot how to fix this problem?

_P.S.: Too sad for the nice test you added just recently!_ :cry:
